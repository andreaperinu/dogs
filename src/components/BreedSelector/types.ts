import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface IBreedSelectorProps {
	maybeSelectedBreed: O.Option<IBreed>,
	maybeBreeds: O.Option<IBreed[]>,
	onChange: (maybeBreed: O.Option<IBreed>) => void
}
