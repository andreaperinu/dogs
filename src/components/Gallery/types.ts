import * as O from 'fp-ts/lib/Option';

export interface IGalleryProps {
	maybeImages: O.Option<string[]>
}
