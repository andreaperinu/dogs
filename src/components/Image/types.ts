import * as O from 'fp-ts/lib/Option'

export interface IImageProps {
	maybeImage: O.Option<string>
	fixedSize?: boolean
}
