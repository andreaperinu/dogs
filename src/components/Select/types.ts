import * as O from 'fp-ts/lib/Option';
import { BaseComponent } from '../../types';

export interface ISelectProps extends BaseComponent {
	maybeValue: O.Option<ISelectItem>
	maybeOptions: O.Option<ISelectItem[]>
	maybeText?: O.Option<string>
	disabled?: boolean
	onChange: (item: O.Option<ISelectItem>) => void
}

export interface ISelectItem {
	id: string
	value: string
}

export interface ISelectStyledProps {
	showOptions: boolean
	disabled: boolean
}
