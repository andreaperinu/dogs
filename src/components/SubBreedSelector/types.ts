import * as O from 'fp-ts/lib/Option';

import { IBreed } from '@/models/models';

export interface ISubBreedSelectorProps {
	disabled?: boolean
	maybeSelectedSubBreed: O.Option<string>,
	maybeBreed: O.Option<IBreed>,
	onChange: (maybeSubBreed: O.Option<string>) => void
}

export interface ISubBreedSelectorStyledProps {
	disabled: boolean
}
