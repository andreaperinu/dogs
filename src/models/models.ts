export interface IBreed {
	breed: string
	subBreeds: string[]
}

export interface IResponse<T> {
	message?: T
	status: string
}

export type IBreedResponseMessage = {
	[breed: string]: string[]
}
