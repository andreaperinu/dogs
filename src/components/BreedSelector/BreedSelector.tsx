import * as O from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';
import { useCallback } from 'react';

import {
	maybeBreedBySelectItem, maybeBreedsToSelectItems, maybeBreedToSelectItem
} from '@/utils/utils';

import { Select } from '../Select/Select';
import { ISelectItem } from '../Select/types';
import { IBreedSelectorProps } from './types';

export const BreedSelector = ({
	maybeSelectedBreed, maybeBreeds, onChange,
}: IBreedSelectorProps) => {
	
	const onChangeHandler = useCallback((maybeSelectItem: O.Option<ISelectItem>) => pipe(
		maybeSelectItem,
		maybeBreedBySelectItem(maybeBreeds),
		onChange,
	), [maybeBreeds])

	return (
		<>
			<h2>Select a breed</h2>
			<Select
				maybeValue={pipe(maybeSelectedBreed, maybeBreedToSelectItem)}
				maybeOptions={pipe(maybeBreeds, maybeBreedsToSelectItems)}
				onChange={onChangeHandler}
				data-test="component-select"
			/>
		</>
	)
}
