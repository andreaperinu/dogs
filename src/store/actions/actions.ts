import {
	FETCH_BREEDS, FETCH_IMAGES, FETCH_RANDOM_IMAGE, RESET_IMAGES, RESET_RANDOM_IMAGE
} from '@/store/actions/rawActions';
import { Action } from '@/store/types';

export type BreedPayload = { breed: string }

export const fetchBreeds: Action = () => ({ type: FETCH_BREEDS })
export const fetchRandomImage: Action<BreedPayload> = (payload) => (
	{ type: FETCH_RANDOM_IMAGE, payload }
)

export const resetRandomImage: Action = () => ({ type: RESET_RANDOM_IMAGE })
export const resetImages: Action = () => ({ type: RESET_IMAGES })

export const fetchImages: Action<BreedPayload> = (payload) => ({ type: FETCH_IMAGES, payload })
