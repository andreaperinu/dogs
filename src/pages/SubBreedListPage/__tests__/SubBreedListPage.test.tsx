import { IBreed } from '@/models/models';
import { setup } from '@/tests/utils';
import { pipe } from 'fp-ts/lib/function';
import * as mockedBreeds from '@/tests/mockedBreeds.json';
import * as A from 'fp-ts/lib/Array'
import component from '../SubBreedListPage';
import * as O from 'fp-ts/lib/Option'
import { formatBreeds } from '@/utils/utils';
import React from 'react'
import { BreedPayload } from '@/store/actions/actions';
import { sequenceT } from 'fp-ts/lib/Apply';
import { IStoreState } from '@/store/types';
import { FETCH_IMAGES, RESET_IMAGES } from '@/store/actions/rawActions';
import { BreedSelector } from '@/components/BreedSelector/BreedSelector';

const dispatch = jest.fn()
const useState = jest.fn()

const maybeSelectedBreed: O.Option<IBreed> = pipe(
	mockedBreeds,
	formatBreeds,
	A.head,
)

const props = {}

const maybeSelectedSubBreed: O.Option<string> = O.some('sub-breed')

const initialState: Partial<IStoreState> = {
	maybeBreeds: O.none,
	maybeImages: O.none,
}

beforeEach(() => {
	useState.mockClear()
	dispatch.mockClear()
})

describe('<SubBreedListPage />', () => {
	it('should match snapshot', () => {
		const wrapper = setup({ component, props, initialState })
		expect(wrapper).toMatchSnapshot()
	})

	it('should dispatch fetch images', () => {
		jest.spyOn(React, 'useState')
			.mockImplementationOnce(() => [maybeSelectedBreed, jest.fn()])
			.mockImplementationOnce(() => [maybeSelectedSubBreed, jest.fn()])

		const payload: BreedPayload = {
			breed: pipe(
				sequenceT(O.option)(maybeSelectedBreed, maybeSelectedSubBreed),
				O.fold(() => '', ([{ breed }, subBreed]) => `${breed}/${subBreed}`)
			)
		}

		setup({ component, props, initialState, dispatch, shouldMount: true, })

		expect(dispatch).toHaveBeenCalledWith({ payload, type: FETCH_IMAGES })
	})

	it('should set breed', () => {
		jest.spyOn(React, 'useState')
			.mockImplementationOnce(init => [init, useState])
			.mockImplementation(init => [init, jest.fn()])

		const wrapper = setup({ component, props, initialState })
		const breedSelector = wrapper.find(BreedSelector)

		breedSelector?.simulate('change', maybeSelectedBreed)

		expect(useState).toHaveBeenCalledWith(maybeSelectedBreed)
	})

	it('should reset sub-breed', () => {
		jest.spyOn(React, 'useState')
			.mockImplementationOnce(init => [init, jest.fn()])
			.mockImplementationOnce(init => [init, useState])

		const wrapper = setup({ component, props, initialState })
		const breedSelector = wrapper.find(BreedSelector)

		breedSelector?.simulate('change', maybeSelectedBreed)

		expect(useState).toHaveBeenCalledWith(O.none)
	})

	it('should dispatch reset random image', () => {
		jest.spyOn(React, 'useState').mockImplementation(init => [init, jest.fn()])

		const wrapper = setup({ component, props, initialState, dispatch })
		const breedSelector = wrapper.find(BreedSelector)

		breedSelector?.simulate('change', maybeSelectedBreed)

		expect(dispatch).toHaveBeenCalledWith({ type: RESET_IMAGES })
	})
})
