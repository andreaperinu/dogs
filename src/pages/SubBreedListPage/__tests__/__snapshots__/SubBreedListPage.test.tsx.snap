// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubBreedListPage /> should match snapshot 1`] = `
<Fragment>
  <BreedSelector
    maybeBreeds={
      Object {
        "_tag": "None",
      }
    }
    maybeSelectedBreed={
      Object {
        "_tag": "None",
      }
    }
    onChange={[Function]}
  />
  <SubBreedSelector
    disabled={true}
    maybeBreed={
      Object {
        "_tag": "None",
      }
    }
    maybeSelectedSubBreed={
      Object {
        "_tag": "None",
      }
    }
    onChange={[Function]}
  />
  <Gallery
    maybeImages={
      Object {
        "_tag": "None",
      }
    }
  />
</Fragment>
`;
